src/datatypes.jl

```diff
@@ -53,11 +136,47 @@ for (mpiname, T) in [
     @eval if @isdefined($(Symbol(:MPI_,mpiname)))
         const $mpiname = _Datatype($(Symbol(:MPI_,mpiname)))
         if !hasmethod(Datatype, Tuple{Type{$T}})
-            Datatype(::Type{$T}; commit=true) = $mpiname
+            Datatype(::Type{$T}) = $mpiname
+            push!(mpi_init_hooks, () -> set_attr!($mpiname, JULIA_TYPE_PTR_ATTR[], pointer_from_objref($T)))
         end
     end
 end
-    
+
+
+@generated function Datatype(::Type{T}) where {T}
+    newtype = Datatype()
+    quote
+        datatype = $newtype
+        # lazily initialize so that it can be safely precompiled
+        if datatype.val === MPI.MPI_Datatype_default
+            Types.create!(datatype, T)
+            Types.commit!(datatype)
+            set_attr!(datatype, JULIA_TYPE_PTR_ATTR[], pointer_from_objref(T))
+        end
+        return datatype
+    end
+end
```

Comment:
@vchuravy Any thoughts on the pros/cons of using a generated function here?

Reply:
hm... this __should__ be okay.

Reply:
YOLO